### ISSUE TITLE
docs: provide `:help lua-plugin`

### ISSUE THREAD

**Issue by glacambre** [neovim/neovim#22366]
### Problem

Some users seem to find the documentation hard to navigate and have a hard time figuring out all the bits and pieces necessary to build Neovim plugins.

### Expected behavior

Neovim could provide a "write lua plugins" guide, which could contain a short description of how plugins are structured (plugin/, autoload/ etc) along with copy-pastable examples showing how to:

- create user-defined commands
- create autocommands
- spawn jobs 
- write plugin documentation
- make colorschemes?

It should not be exhaustive in order to remain easily consumed.

**Comment by bew** [comment]
It would be interesting to add a kind of faq for when we're writing a plugin. Or more like a general guide of `recommended ways to do X or Y high-level thing`

For example I recently asked myself `I want to give an error message to the user`, print is nice for debugging, but there's no real help/links to other things that might be better. Redirecting to sth like `vim.api.nvim_err_writeln` would help.
It could also explain how to 'raise' Lua errors from low-level core functions, that are later transformed to friendly errors.
Or ``I want to expose map-able actions (not necessarily with config in a `setup` function)``, talk about actions in different modes, actions that return an expr, why / when to choose which..
(And I'm sure there are many others)

**Comment by mfussenegger** [comment]
> could contain a short description of how plugins are structured (plugin/, autoload/ etc)

That probably opens up a general question of whether Lua content should be side-by-side to existing vimscript content - potentially duplicating a lot of information. Or if the existing content should be updated to include Lua. There is `:h plugin` and `:h write-plugin`. A lot of that content applies to Lua as well.

**Comment by justinmk** [comment]
> There is `:h plugin` and `:h write-plugin`. A lot of that content applies to Lua as well.

Yep. Writing a user manual is out of scope for this issue, and imo out of scope for in general, for now. 

`:help lua-guide` is the Nvim "user manual", for Lua topics.

**Comment by dundargoc** [comment]
As neovim's Chief Documentation Officer, here are my thoughts on the given topics:

- "create user-defined commands" and "create autocommands": I feel like these are adequately covered in `lua-guide`, which is where I think they should belong. After all, these aren't specific to plugins but very much relevant to user configurations as well. If something is missing I'd rather see these fixed in the `lua-guide`.

- "spawn jobs": I think out of all the points given, this one has the best potential to result in something useful. Bonus points for being neovim-specific rather than "lua plugins are like vimscript plugins but in lua". I'm not entirely convinced this alone is enough to make up for an entire plugin guide. I think a polishing of `channel.txt` could work just as well.

- "write plugin documentation": I think this is purely a vim issue? This is addressed by `:h write-local-help` and `:h add-local-help`. I suppose we could aggregate this information in a more presentable manner but I don't think repackaging vim documentation like that is a good use of our time. This can be improved on the Vim side which we can then import. Bonus points as both projects benefit.

- "make colorschemes?": this one is out of my depth, mostly because I haven't bothered to understand how colors are made. I will skip commenting on this.

All in all, I'm not convinced these are enough justification for a dedicated lua-plugin-guide. The nice thing about lua-guide is that its scope is fairly limited: "here are the basics you need to write a simple neovim config in lua, good luck". I have a feeling a lua-plugin-guide can blow up in scope and become unfocused as plugins can vary wildly between each other.

Another reservation I have with this is purely practical: nanotee had already done the hard part for us by curating the needed topics for lua-guide and it *still* took an obscene amount of time to rewrite and review. I'm dreading writing this type of guide from scratch and fear it may be an absolute time sink.

I'm purely speculating here, but I wonder if the best way to make this a reality isn't for a dedicated person to champion this in their own repo, just like nanotee did? That way the guide has time to mature, learn from its mistakes and get to polish over time. I think it'd also be a good way to gague the actual interest for something like this.

**Comment by justinmk** [comment]
# Proposal

A very small, copy-paste boilerplate for creating a new plugin will solve 80% of the problem. Maybe times, the friction is just in knowing where the file goes and how to get it started, plus some sample functions. Look at `:help health-dev` (50 lines!) for what I am imagining.

**Comment by lewis6991** [comment]
In addition, to combat the "must lazy load everything" mentality that has developed in the community, such docs could also outline best practices for structuring a plugin.

E.g.:

- Keep anything in `/plugin` light. Use it to mainly define commands and autocmds.
- Defer `require` as much as possible.
- Provide a checkhealth entry.
- Make use of `vim.deprecate`.
